src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -269,48 +294,99 @@ public TypeSymbolWithAnnotations SetIsAnnotated(CSharpCompilation compilation)
         /// </summary>
-        private static bool? MergeIsNullable(bool? a, bool? b, VarianceKind variance, out bool hadNullabilityMismatch)
+        private static NullableAnnotation MergeNullableAnnotation(TypeSymbol type, NullableAnnotation a, NullableAnnotation b, VarianceKind variance, out bool hadNullabilityMismatch)
         {
             hadNullabilityMismatch = false;
             if (a == b)
             {
                 return a;
             }
+
+            if (a.IsAnyNullable() && b.IsAnyNullable())
+            {
+                return NullableAnnotation.Nullable;
+            }
+
             switch (variance)
             {
                 case VarianceKind.In:
-                    return (a == false || b == false) ? (bool?)false : null;
+
+                    if (a == NullableAnnotation.NotNullableBasedOnAnalysis || b == NullableAnnotation.NotNullableBasedOnAnalysis)
+                    {
+                        return NullableAnnotation.NotNullableBasedOnAnalysis;
+                    }
+
+                    if (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable)
+                    {
+                        return NullableAnnotation.NotNullable;
+                    }
+
+                    Debug.Assert(a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown);
+                    return NullableAnnotation.Unknown;
+
                 case VarianceKind.Out:
-                    return (a == true || b == true) ? (bool?)true : null;
+                    if (a.IsAnyNullable())
+                    {
+                        Debug.Assert(!b.IsAnyNullable());
+                        return a;
+                    }
+
+                    if (b.IsAnyNullable())
+                    {
+                        return b;
+                    }
+
+                    if (a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown)
+                    {
+                        return NullableAnnotation.Unknown;
+                    }
+
+                    Debug.Assert((a == NullableAnnotation.NotNullable && b == NullableAnnotation.NotNullableBasedOnAnalysis) ||
+                        (b == NullableAnnotation.NotNullable && a == NullableAnnotation.NotNullableBasedOnAnalysis));
+                    return NullableAnnotation.NotNullable;
+
                 default:
-                    if (a == null)
+                    if (a == NullableAnnotation.Unknown)
                     {
                         return b;
                     }
-                    if (b == null)
+                    if (b == NullableAnnotation.Unknown)
                     {
                         return a;
                     }
+
+                    if ((a == NullableAnnotation.NotNullable && b == NullableAnnotation.NotNullableBasedOnAnalysis) ||
+                        (b == NullableAnnotation.NotNullable && a == NullableAnnotation.NotNullableBasedOnAnalysis))
```

Comment:
In an effort to finish this review, I pushed myself to finish reviewing this method, despite my earlier feedback on clarity. Still, I didn't manage to understand this block.

Why are we looking at `NotNullable` and `NotNullableBasedOnAnalysis` specifically, rather than using broader categories, like `IsAnyNullable`/`IsAnyNotNullable` (or even better using something like `ValueCanBeNull`)?

Consider the following code:
```C#
bool? aIsNullable = ValueCanBeNull(a, type); // static helper method (adding 4 lines to this file)
bool? bIsNullable = ValueCanBeNull(b, type);
switch (variance)
{
...
default:
  if (aIsNullable == false || bIsNullable == false)
  {
    if (asIsNullable == true || aIsNullable == true)
    {
      hadNullabilityMismatch = true;
      return NullableAnnotation.Unknown; // temporarily replicate bug in current implementation, which @gafter already fixed in another PR. Remove this line in final implementation.
    }
    return NullableAnnotation.NotNullableBasedOnAnalysis;
  }

  if (aIsNullable == true || bIsNullable == true)
  {
    return NullableAnnotation.NullableBasedOnAnalysis;
  }

  return NullableAnnotation.Unknown;
}
```

Am I missing something?
If we're having trouble convincing ourselves that such code is correct, then I'm afraid it is even harder to convince ourselves that the current code is correct.

Reply:
> Am I missing something?

Let's take this offline.

---
In reply to: [231718701](https://github.com/dotnet/roslyn/pull/30913#discussion_r231718701) [](ancestors = 231718701)

Reply:
From our discussion, Aleksey clarified that there is an additional requirement: we should return an annotation that existed in the inputs, and we prefer `NotNullable` over `NotNullableBasedOnAnalysis`.

The above code can be modified with those requirements, and for what it's worth the tests pass.
Those requirements do add complexity, because `CanValueBeNull` is lossy, as Aleksey pointed out, so we need to do extra work to return a value from the inputs.

```C#
// The common code above the switch can be moved into the other cases above, or can be left as-is.
                default:

                    bool? aIsNullable = ValueCanBeNull(a, type);
                    bool? bIsNullable = ValueCanBeNull(b, type);
                    if (aIsNullable == false || bIsNullable == false)
                    {
                        if (aIsNullable == true || bIsNullable == true)
                        {
                            hadNullabilityMismatch = true;
                            return NullableAnnotation.Unknown; // temporarily replicate bug in current implementation, which @gafter already fixed in another PR. Remove this line in final implementation.
                        }

                        // prefer NotNullable to NotNullableBasedOnAnalysis
                        return (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable) ? 
                            NullableAnnotation.NotNullable : NullableAnnotation.NotNullableBasedOnAnalysis;
                    }

                    if (aIsNullable == true || bIsNullable == true)
                    {
                        // return annotation from input
                        return a.IsAnyNullable() ? a : b;
                    }

                    return NullableAnnotation.Unknown;
 ```